ci: enable CodeQL code scanning

matifali opened the pull request:
<!--
Check if your change requires documentation edits before merging: https://coder.com/docs/coder. Make edits in `docs/`.
-->

matifali committed:
chore: enable GraphQL code scanning

matifali commented:
Hi @kylecarbs. What do you think about running this security analysis?

matifali committed:
Merge branch 'coder:main' into GraphQL-code-scanning

matifali commented:
@ghuntley. You may close this if not planned.

mafredri commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,74 @@
+
+    # - run: |
+    #   echo "Run, Build Application using script"
+    #   ./location_of_script_within_repo/buildscript.sh
```
Should we build the project via Makefile here? I'm guessing scanning the binaries would also be beneficial? (I'm guessing for e.g. JS we only build the site package.)

deansheather replied:
We don't need to build the Go binary. It just needs the dependencies downloaded, so we added `go mod download` as a step

mafredri commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,74 @@
+    strategy:
+      fail-fast: false
+      matrix:
+        language: [ 'go', 'javascript' ]
```
Should we limit JavaScript scope to `./site`, exclude `node_modules`? (https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-code-scanning#specifying-directories-to-scan)

matifali replied:
Yes we may limit this scanning to only `./site`

mafredri reviewed (commented):
Personally I think this could be a nice addition, thanks for the PR! I'd like to know what @kylecarbs and @ghuntley think though (maybe @deansheather too?). And if we do want to merge, we should fix/silence the current errors first.

(PS. To fix the test/js failure, rebase/merge on top of master.)

deansheather commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,74 @@
+      fail-fast: false
+      matrix:
+        language: [ 'go', 'javascript' ]
+        # CodeQL supports [ 'cpp', 'csharp', 'go', 'java', 'javascript', 'python', 'ruby' ]
+        # Learn more about CodeQL language support at https://aka.ms/codeql-docs/language-support
```
```suggestion
```

matifali committed:
removed the top commnets and description.

Co-authored-by: Dean Sheather <dean@deansheather.com>

matifali committed:
Update .github/workflows/codeql.yml

Co-authored-by: Dean Sheather <dean@deansheather.com>

matifali committed:
Update .github/workflows/codeql.yml

Co-authored-by: Dean Sheather <dean@deansheather.com>

matifali committed:
Update .github/workflows/codeql.yml

Co-authored-by: Dean Sheather <dean@deansheather.com>

matifali committed:
Update .github/workflows/codeql.yml

Co-authored-by: Dean Sheather <dean@deansheather.com>

matifali committed:
Comment on when the action schedule runs

matifali committed:
Merge branch 'coder:main' into GraphQL-code-scanning

deansheather committed:
Update .github/workflows/codeql.yml

deansheather commented:
I think this is good to merge but I would like approval from @ElliotG first. I will ask them in slack

deansheather commented:
Elliot is unavailable at the moment so once Mathias approves we can merge.

mafredri commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,43 @@
+      with:
+        languages: ${{ matrix.language }}
+
+    - name: Download Go Dependencies
+      if: matrix.language == 'go'
+      run: go mod download
```
Based on output from https://github.com/coder/coder/actions/runs/3737608156/jobs/6342956567, it looks like Go isn't available? Go mod download results in no output. The [Python example](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-code-scanning#analyzing-python-dependencies) does manual setup of Python, so maybe it will help here.

Also, it looks like "Perform Analysis" tries to build a bunch of stuff, including the node/site package and takes a whole 15 minutes to complete. Maybe mod download failure is the reason, or we need another way to disable the autobuilding.

```suggestion
    - name: Setup Go
      if: matrix.language == 'go'
      uses: actions/setup-go@v3
      with:
        go-version: "~1.19"

    - name: Download Go Dependencies
      if: matrix.language == 'go'
      run: go mod download
```

mafredri replied:
Hmm, looks like this didn't change the analysis step. Maybe the step works, but analysis still triggers a `go mod download` via build-step as well. I'm confused as to why it's behaving that way.

matifali committed:
Update .github/workflows/codeql.yml

Co-authored-by: Mathias Fredriksson <mafredri@gmail.com>

mafredri commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,49 @@
+      with:
+        languages: ${{ matrix.language }}
+
+    - name: Setup Go
+      if: matrix.language == 'go'
+      uses: actions/setup-go@v3
+      with:
+        go-version: "~1.19"
+
+    - name: Download Go Dependencies
+      if: matrix.language == 'go'
+      run: go mod download
```
```suggestion
    - name: Download Go Dependencies
      if: matrix.language == 'go'
      run: |
        go mod download
        # Disable Analysis step from trying to build the project.
        rm Makefile
```

mafredri committed:
Update .github/workflows/codeql.yml

mafredri commented:
With https://github.com/coder/coder/commit/bcde5f3fc5f0e1b4eb772bd39a898bb0348ef6a9 it now runs faster (still pretty slow @ 7+ minutes though). It's a bit of an ugly hack but it works.

The `go mod download` seems useless though, the perform analysis step does it's own download/module install even though we do the download beforehand.

mafredri commented:
Actually, we may want to put back the manual installation of specific Go version, just spotted this output:

```
2022/12/20 11:10:43 Autobuilder was built with go1.19.3, environment has go1.18.9
[52](https://github.com/coder/coder/actions/runs/3739932860/jobs/6347716302#step:5:53)
  2022/12/20 11:10:43 LGTM_SRC is /home/runner/work/coder/coder
[53](https://github.com/coder/coder/actions/runs/3739932860/jobs/6347716302#step:5:54)
  2022/12/20 11:10:43 Found go.mod, enabling go modules
[54](https://github.com/coder/coder/actions/runs/3739932860/jobs/6347716302#step:5:55)
  go: go.mod file indicates go 1.19, but maximum version supported by tidy is 1.18
[55](https://github.com/coder/coder/actions/runs/3739932860/jobs/6347716302#step:5:56)
```

Since we use go 1.19, I'm guessing it'd be good for the environment to be go 1.19 as well.

matifali committed:
Update codeql.yml

Updated as per @mafredri suggestion.

matifali commented:
> Actually, we may want to put back the manual installation of specific Go version, just spotted this output:
> 
> ```
> 2022/12/20 11:10:43 Autobuilder was built with go1.19.3, environment has go1.18.9
> [52](https://github.com/coder/coder/actions/runs/3739932860/jobs/6347716302#step:5:53)
>   2022/12/20 11:10:43 LGTM_SRC is /home/runner/work/coder/coder
> [53](https://github.com/coder/coder/actions/runs/3739932860/jobs/6347716302#step:5:54)
>   2022/12/20 11:10:43 Found go.mod, enabling go modules
> [54](https://github.com/coder/coder/actions/runs/3739932860/jobs/6347716302#step:5:55)
>   go: go.mod file indicates go 1.19, but maximum version supported by tidy is 1.18
> [55](https://github.com/coder/coder/actions/runs/3739932860/jobs/6347716302#step:5:56)
> ```
> 
> Since we use go 1.19, I'm guessing it'd be good for the environment to be go 1.19 as well.

Removed `go mod download` and added the `go 1.19` installation back.

matifali committed:
Merge branch 'coder:main' into GraphQL-code-scanning

mafredri commented:
> Removed `go mod download` and added the `go 1.19` installation back.

👍

I noticed the `rm Makefile` also got removed. Without it the analysis step downloads node modules and builds the `site` package (and other make commands as well). It adds unneeded execution time. I would love another way to disable it though but found nothing in the docs.

ghuntley committed:
add vscode extension recommendation

matifali committed:
remove MakeFile

matifali committed:
Merge branch 'coder:main' into GraphQL-code-scanning

matifali commented:
> > Removed `go mod download` and added the `go 1.19` installation back.
> 
> 
> 
> 👍
> 
> 
> 
> I noticed the `rm Makefile` also got removed. Without it the analysis step (for Go) downloads node modules and builds the `site` package (and other make commands as well). It adds unneeded execution time. I would love another way to disable it though but found nothing in the docs.

Added that back

mafredri committed:
Enable go mod cache

mafredri committed:
fmt

mafredri commented:
> Added that back @mafredri But still taking +7 minutes.

Thanks. Yeah it's too bad. I'm not sure if we can beat the 7min by much but I tried adding go mod caching in hopes that it'd save some time, let's see.

mafredri commented:
Thought it’d be nice to get this in so I hit merge. Thanks for your contribution @matifali! (We can always iterate on this further if something needs tweaking.)
